utils/unittest/entity.go

```diff
@@ -0,0 +1,245 @@
+	return func(t *MalleabilityChecker) {
+		t.customTypes[reflect.TypeOf(tType)] = func() reflect.Value {
+			return reflect.ValueOf(generator())
+		}
+	}
+}
+
+// NewMalleabilityChecker creates a new instance of the MalleabilityChecker with the given options.
+func NewMalleabilityChecker(t *testing.T, ops ...MalleabilityCheckerOpt) *MalleabilityChecker {
+	checker := &MalleabilityChecker{
+		T:           t,
+		customTypes: make(map[reflect.Type]func() reflect.Value),
+	}
+	for _, op := range ops {
+		op(checker)
+	}
+	return checker
+}
+
+// Check is a method that performs the malleability check on the entity.
+func (t *MalleabilityChecker) Check(entity flow.IDEntity) {
+	v := reflect.ValueOf(entity)
+	if v.Kind() == reflect.Ptr {
+		require.False(t, v.IsNil(), "entity is nil, nothing to check")
+		v = v.Elem()
+	}
+	isMalleable := t.isEntityMalleable(v, entity.ID)
+	require.False(t, isMalleable, "entity is malleable")
+}
+
+// isEntityMalleable is a helper function to recursively check fields of the entity. Every time we change a field we check if ID of the entity has changed.
+// If ID has not changed then entity is malleable.
+// This function returns boolean value so we can add extra traces to the test output in case of recursive calls to structure fields.
+func (t *MalleabilityChecker) isEntityMalleable(v reflect.Value, idFunc func() flow.Identifier) bool {
+	if v.Kind() == reflect.Ptr && v.IsNil() {
+		return false // there is nothing to check for this field if it's nil
+	}
+
+	tType := v.Type()
+	// if we have a custom type function we should use it to generate a random value for the field.
+	customTypeGenerator, hasCustomTypeOverride := t.customTypes[tType]
+	if hasCustomTypeOverride {
+		origID := idFunc()
+		v.Set(customTypeGenerator())
+		return origID == idFunc()
+	}
+
+	if v.Kind() == reflect.Struct {
+		// in case of a struct we would like to ensure that changing any field will change the ID of the entity.
+		// we will recursively check all fields of the struct and if any of the fields is malleable then the whole entity is malleable.
+		for i := 0; i < v.NumField(); i++ {
+			field := v.Field(i)
+			require.Truef(t, field.CanSet(), "field %s is not settable", tType.Field(i).Name)
+			require.False(t, t.isEntityMalleable(field, idFunc), "field %s is malleable", tType.Field(i).Name)
+		}
+		return false
+	} else {
+		// when dealing with non-composite type we can generate random values for it and check if ID has changed.
+		origID := idFunc()
+		expectChange := generateRandomReflectValue(t.T, v)
+		newID := idFunc()
+		return expectChange && origID == newID
+	}
+}
+
+// generateRandomReflectValue uses reflection to switch on the field type and generate a random value for it.
+func generateRandomReflectValue(t *testing.T, field reflect.Value) bool {
+	generated := true
+	switch field.Kind() {
+	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
+		field.SetUint(field.Uint() + uint64(rand.Intn(100)+1))
```

Comment:
We could even use `^field.Uint()` to make sure we're touching every bit of the integer, to guarantee that ID() involves all bits of the integer fields. That should cover more cases than randomness with a cheap cost

Reply:
great idea